pkg/oauth/tokenrequest/negotiator_gssapi.go

```diff
@@ -0,0 +1,198 @@
+	lib, err := g.loadLib()
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialize our context if we haven't already
+	if g.ctx == nil {
+
+		if len(g.principalName) > 0 {
+			// Get credentials for a specific principal
+			klog.V(5).Infof("acquiring credentials for principal name %s", g.principalName)
+			credBuffer, err := lib.MakeBufferString(g.principalName)
+			if err != nil {
+				return nil, convertGSSAPIError(err)
+			}
+			defer credBuffer.Release()
+
+			credName, err := credBuffer.Name(lib.GSS_KRB5_NT_PRINCIPAL_NAME)
+			if err != nil {
+				return nil, convertGSSAPIError(err)
+			}
+			defer credName.Release()
+
+			cred, _, _, err := lib.AcquireCred(credName, time.Duration(0), lib.GSS_C_NO_OID_SET, gssapi.GSS_C_INITIATE)
+			if err != nil {
+				klog.V(5).Infof("AcquireCred returned error: %v", err)
+				return nil, convertGSSAPIError(err)
+			}
+			g.cred = cred
+		} else {
+			// otherwise, express no opinion about the credentials and let gssapi decide
+			g.cred = lib.GSS_C_NO_CREDENTIAL
+		}
+
+		serviceName, err := getServiceName('@', requestURL)
+		if err != nil {
+			return nil, err
+		}
+
+		klog.V(5).Infof("importing service name %s", serviceName)
+		nameBuf, err := lib.MakeBufferString(serviceName)
+		if err != nil {
+			return nil, convertGSSAPIError(err)
+		}
+		defer nameBuf.Release()
+
+		name, err := nameBuf.Name(lib.GSS_C_NT_HOSTBASED_SERVICE)
+		if err != nil {
+			return nil, convertGSSAPIError(err)
+		}
+		g.name = name
+		g.ctx = lib.GSS_C_NO_CONTEXT
+	}
+
+	incomingTokenBuffer, err := lib.MakeBufferBytes(challengeToken)
+	if err != nil {
+		return nil, convertGSSAPIError(err)
+	}
+	defer incomingTokenBuffer.Release()
+
+	var outgoingToken *gssapi.Buffer
+	g.ctx, _, outgoingToken, _, _, err = lib.InitSecContext(g.cred, g.ctx, g.name, lib.GSS_C_NO_OID, g.flags, time.Duration(0), lib.GSS_C_NO_CHANNEL_BINDINGS, incomingTokenBuffer)
+	defer outgoingToken.Release()
+
+	switch err {
+	case nil:
+		klog.V(5).Infof("InitSecContext returned GSS_S_COMPLETE")
+		g.complete = true
+		return outgoingToken.Bytes(), nil
+	case gssapi.ErrContinueNeeded:
+		klog.V(5).Infof("InitSecContext returned GSS_S_CONTINUE_NEEDED")
+		g.complete = false
+		return outgoingToken.Bytes(), nil
+	default:
+		klog.V(5).Infof("InitSecContext returned error: %v", err)
+		return nil, convertGSSAPIError(err)
+	}
+}
+
+func (g *gssapiNegotiator) IsComplete() bool {
+	return g.complete
+}
+
+func (g *gssapiNegotiator) Release() error {
+	var errs []error
+	if err := g.name.Release(); err != nil {
+		errs = append(errs, convertGSSAPIError(err))
+	}
+	if err := g.ctx.Release(); err != nil {
+		errs = append(errs, convertGSSAPIError(err))
+	}
+	if err := g.cred.Release(); err != nil {
+		errs = append(errs, convertGSSAPIError(err))
+	}
+	if err := g.lib.Unload(); err != nil {
+		errs = append(errs, convertGSSAPIError(err))
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+func (g *gssapiNegotiator) loadLib() (*gssapi.Lib, error) {
```

Comment:
oh dear :see_no_evil:

Reply:
yeah, that's the bit I told you we should investigate long-term to replace with a go-based solution